### PR TITLE
Fix log batch schema validation

### PR DESCRIPTION
**Pull request by mrkaye97**
## Related Issues/PRs

#5458 

Fixes an issue with required parameters in `/log-batch`. There were a few problems:
1. `step` was being asserted as required by the API when it shouldn't be required (per the docs)
2. If a key was missing from the metrics dict, then `metrics[<< missing key >>]` would error and the request would `500` before being able to get to the assertions that would otherwise `400`.

## How is this patch tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests (describe details, including test results, below)

A couple example calls from R:

### Old (broken) version (returns `500` with a bunch of HTML content) running against our prod MLFlow server (v2.3.0):

`500` if timestamp isn't specified

```r
body <- jsonlite::fromJSON('{
   "run_id": "2c5b5449bfb04a489c61c3617f522e03",
   "metrics": [
     {"key": "mae", "value": 2.5, "step": 1}
   ]
}')

> POST(
+   "https://<snip>/api/2.0/mlflow/runs/log-batch",
+   body = jsonlite::toJSON(body, pretty = TRUE, auto_unbox = TRUE),
+   add_headers(
+     c(
+       Authorization = auth_header
+     )
+   )
+ ) %>%
+   content()
{html_document}
<html lang="en">
[1] <head>\n<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">\n<title>50 ...
[2] <body>\n<h1>Internal Server Error</h1>\n<p>The server encountered an internal error an ...
```

`500` if step isn't specified:

```r
> body <- jsonlite::fromJSON('{
+    "run_id": "2c5b5449bfb04a489c61c3617f522e03",
+    "metrics": [
+      {"key": "mae", "value": 2.5, "timestamp": 1}
+    ]
+ }')
> POST(
+   "https://<snip>/api/2.0/mlflow/runs/log-batch",
+   body = jsonlite::toJSON(body, pretty = TRUE, auto_unbox = TRUE),
+   add_headers(
+     c(
+       Authorization = auth_header
+     )
+   )
+ ) %>%
+   content()
{html_document}
<html lang="en">
[1] <head>\n<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">\n<title>50 ...
[2] <body>\n<h1>Internal Server Error</h1>\n<p>The server encountered an internal error an ...
```

here's what I see in the logs:

```
2023-04-28T19:23:09.960261+00:00 app[web.1]: 2023/04/28 19:23:09 ERROR mlflow.server: Exception on /api/2.0/mlflow/runs/log-batch [POST]
2023-04-28T19:23:09.960281+00:00 app[web.1]: Traceback (most recent call last):
2023-04-28T19:23:09.960281+00:00 app[web.1]:   File "/opt/conda/lib/python3.10/site-packages/flask/app.py", line 2190, in wsgi_app
2023-04-28T19:23:09.960281+00:00 app[web.1]:     response = self.full_dispatch_request()
2023-04-28T19:23:09.960282+00:00 app[web.1]:   File "/opt/conda/lib/python3.10/site-packages/flask/app.py", line 1486, in full_dispatch_request
2023-04-28T19:23:09.960282+00:00 app[web.1]:     rv = self.handle_user_exception(e)
2023-04-28T19:23:09.960283+00:00 app[web.1]:   File "/opt/conda/lib/python3.10/site-packages/flask/app.py", line 1484, in full_dispatch_request
2023-04-28T19:23:09.960283+00:00 app[web.1]:     rv = self.dispatch_request()
2023-04-28T19:23:09.960283+00:00 app[web.1]:   File "/opt/conda/lib/python3.10/site-packages/flask/app.py", line 1469, in dispatch_request
2023-04-28T19:23:09.960283+00:00 app[web.1]:     return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
2023-04-28T19:23:09.960283+00:00 app[web.1]:   File "/opt/conda/lib/python3.10/site-packages/mlflow/server/handlers.py", line 475, in wrapper
2023-04-28T19:23:09.960284+00:00 app[web.1]:     return func(*args, **kwargs)
2023-04-28T19:23:09.960284+00:00 app[web.1]:   File "/opt/conda/lib/python3.10/site-packages/mlflow/server/handlers.py", line 516, in wrapper
2023-04-28T19:23:09.960284+00:00 app[web.1]:     return func(*args, **kwargs)
2023-04-28T19:23:09.960285+00:00 app[web.1]:   File "/opt/conda/lib/python3.10/site-packages/mlflow/server/handlers.py", line 1108, in _log_batch
2023-04-28T19:23:09.960285+00:00 app[web.1]:     request_message = _get_request_message(
2023-04-28T19:23:09.960285+00:00 app[web.1]:   File "/opt/conda/lib/python3.10/site-packages/mlflow/server/handlers.py", line 440, in _get_request_message
2023-04-28T19:23:09.960285+00:00 app[web.1]:     _validate_param_against_schema(
2023-04-28T19:23:09.960285+00:00 app[web.1]:   File "/opt/conda/lib/python3.10/site-packages/mlflow/server/handlers.py", line 365, in _validate_param_against_schema
2023-04-28T19:23:09.960286+00:00 app[web.1]:     f(value)
2023-04-28T19:23:09.960286+00:00 app[web.1]:   File "/opt/conda/lib/python3.10/site-packages/mlflow/server/handlers.py", line 1100, in _assert_metrics_fields_present
2023-04-28T19:23:09.960286+00:00 app[web.1]:     _assert_required(m["timestamp"])
2023-04-28T19:23:09.960287+00:00 app[web.1]: KeyError: 'timestamp'
```

### New (working) version w/ this change:

```r
> body <- jsonlite::fromJSON('{
+    "run_id": "2c5b5449bfb04a489c61c3617f522e03",
+    "metrics": [
+      {"key": "mae", "value": 2.5, "timestamp": 1}
+    ]
+ }')

> POST(
+   "http://127.0.0.1:5000/api/2.0/mlflow/runs/log-batch",
+   body = jsonlite::toJSON(body, pretty = TRUE, auto_unbox = TRUE)
+ ) %>%
+   status_code()
[1] 200
```

Ideally, this would tell the user that `timestamp` was missing (the hint is pretty misleading in this case) but that's just not how the validation is implemented, and I didn't want to change the guts of the operation too much.

### New version, missing timestamp correctly `400`s:
```r
> body <- jsonlite::fromJSON('{
+    "run_id": "2c5b5449bfb04a489c61c3617f522e03",
+    "metrics": [
+      {"key": "mae", "value": 2.5, "step": 1}
+    ]
+ }')

> POST(
+   "http://127.0.0.1:5000/api/2.0/mlflow/runs/log-batch",
+   body = jsonlite::toJSON(body, pretty = TRUE, auto_unbox = TRUE)
+ ) %>%
+   content()
$error_code
[1] "INVALID_PARAMETER_VALUE"

$message
[1] "Missing value for required parameter 'metrics'. See the API docs for more information about request parameters."
```

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [x] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/server/handlers.py [hunk 1]
```diff
@@ -1095,14 +1094,13 @@ def _get_artifact_repo(run):
 def _log_batch():
     def _assert_metrics_fields_present(metrics):
         for m in metrics:
-            _assert_required(m["key"])
-            _assert_required(m["value"])
-            _assert_required(m["timestamp"])
-            _assert_required(m["step"])
```
Comment (mrkaye97) [inline mlflow/server/handlers.py:1101]: note that `step` is now removed as a required param, per the docs

File: tests/tracking/test_rest_tracking.py [hunk 2]
```diff
@@ -673,6 +673,31 @@ def assert_bad_request(payload, expected_error_message):
...
 
+    ## Should 400 if missing timestamp
+    assert_bad_request(
+        {
+            "run_id": run_id,
+            "metrics": [
+                {"key": "mae", "value": 2.5}
+            ]
+        },
+        'Invalid value [{"key": "mae", "value": 2.5}] for parameter \'metrics\' supplied'
```
Comment (mrkaye97) [inline tests/tracking/test_rest_tracking.py:684]: not that familiar with the test suite -- is there a way to remove the list of values and replace it with a `*` or something to be agnostic to the actual `metrics` list in the request?

**Comment by mrkaye97** [comment]
Also am not entirely sure what the linter is mad about. I tried following the docs in `CONTRIBUTING.md` saying to run `./dev/lint.sh`, but that file doesn't exist

**Comment by dbczumar** [comment]
@mlflow-automation autoformat

**Review by dbczumar — APPROVED** [review]
LGTM! Thanks @mrkaye97 !

**Comment by dbczumar** [comment]
Lint failures are unrelated and will be fixed in a separate PR. Merging :)

**Pull request merged**